### DISCUSSION TITLE
Autoscaling with k8s

### DISCUSSION THREAD

**Discussion by jsoref** [ChristopherHX/github-act-runner#52] [Ideas]
I understand this might be a bit outside of scope for github-act-runner, but I suspect that the work you've done might have some of the pieces necessary to do it...

Currently, we can register fixed runners to GitHub, but each runner has to be provisioned against actual hardware (even if it's virtualized), which means we're billed for its idle time.

We'd love to be able to instantiate a runner-operator which registers a proxy for a runner with github, and when it gets a job, asks k8s to create a pod, and then sends the work to the pod when the pod is ready. Essentially, the operator would act as a MITM between the github server and the runner implementation.

When the operator assigns work to a runner, it should check to see if it is configured to allow more concurrent instances, in which case, it should register itself again to github, and wait for new work from that additional connection.

Ideally, it would shut down k8s nodes after some idle period.

I think as I've described, it could probably skip deregistering pool capacity to github, or could even register its full capacity to github up front, so long as it doesn't instantiate the actual pods that do the running until there's work for them.

I've wanted to do this for a couple of months and determined that I'd have to reverse engineer a significant portion of the messaging between a client and the server. I believe this project has already done that.

**Comment by ChristopherHX** [comment]
Do you want one runner-operator
- which long polls a github runner instance
- if it receives a job -> delegate it to a worker which will be spawned via k8s, start to long poll another github runner instance (to receive more jobs)

This requires at least one idle operator / runner with a persistent connection to github.
At least I plan to make executing the worker configurable via some javascript code, so this might be possible at some point of time.

Do you want to use this projects runner (incl. a quite big fork of act for step execution), the official runner or both?

Althought it might be possible pipe the request to an official runner via my other project https://github.com/christopherhx/runner.server (the protocol of this project is written against runner.server, then patched remaining incompatibility with github.com) or calling Runner.Worker directly, without direct registration against github.

For the official runner there exists something to autoscale for k8s https://github.com/actions-runner-controller/actions-runner-controller, I haven't tested this, but who knows how well it works for your usecase.
It seems to be able to scale up via webhooks, which doesn't require a persistent connection to github.

**Reply by jsoref** [reply]
> Do you want one runner-operator
> * which long polls a github runner instance
> * if it receives a job -> delegate it to a worker which will be spawned via k8s, start to long poll another github runner instance (to receive more jobs)

> This requires at least one idle operator / runner with a persistent connection to github.

My goal is to have an idle operator, but not necessarily _any_ idle runners.

> Do you want to use this projects runner (incl. a quite big fork of act for step execution), the official runner or both?

No opinion.

> Althought it might be possible pipe the request to an official runner via my other project https://github.com/christopherhx/runner.server (the protocol of this project is written against runner.server, then patched remaining incompatibility with github.com) or calling Runner.Worker directly, without direct registration against github.

Sounds interesting.

> For the official runner there exists something to autoscale for k8s https://github.com/actions-runner-controller/actions-runner-controller, I haven't tested this, but who knows how well it works for your usecase.
It seems to be able to scale up via webhooks, which doesn't require a persistent connection to github.

Hmm, I'll have to look into this as well.

**Reply by ChristopherHX** [reply]
> could even register its full capacity to github up front, so long as it doesn't instantiate the actual pods that do the running until there's work for them.

Yes adding a configure command to auto register a constant amount of similar runners could be easily integrated without the need to create a github app or store a Personal Access Token. (setting up a github app is a mess compared to registering a runner)
The only thing this operator needs to do is to contact github once per 30 days, otherwise github deregister such an offline instance

I have to look in splitting the code of the listener and the actual worker which runs steps, so the listener (currently called runner) is able to spawn a worker optionally with a custom script where you like e.g. k8s, a vm or even azure and aws.

How are you able to pass a json file to a remote process?
I think a file would always work, I'm not shure about pipes.

The operator need to
- send a json job request to a spawned worker (this message contains a job scoped access token to the github actions service, so the worker cannot receive any new jobs)
- be notified of worker exit (success and failure)
- be able to send the worker a message to cancel the run
  - Would stdin / stdout work to communicate with a k8s instance?

**Reply by johnandersen777** [reply]
Ive had success with adding this repo as the last layer of a container and making the CMD a giant string passed to bash with the config remove (not sure how well this works) as an EXIT trap. Then deploy as a deamomset. This doesn’t solve the auto/scale to zero problem but you could mimic the ARC workflow_job webhook receiver with a service account token in that pod. Id also probably say itd be good to have that help with removal because ive experienced that trap not always running. Theres also a new beta `gh webhook` or something way to get events without having to expose the webbook receiver pod to anything but the local network. One could run that as a sidecar. Also ACT uses the runner hooks to properly run container based actions in their own pods. In my experience only composite and javascript actions work with the deployment methodology ive tried.
